tracing-subscriber/src/layer/layered.rs

```diff
@@ -0,0 +1,536 @@
+        self.inner.on_record(span, values, ctx.clone());
+        self.layer.on_record(span, values, ctx);
+    }
+
+    #[inline]
+    fn on_follows_from(&self, span: &span::Id, follows: &span::Id, ctx: Context<'_, S>) {
+        self.inner.on_follows_from(span, follows, ctx.clone());
+        self.layer.on_follows_from(span, follows, ctx);
+    }
+
+    #[inline]
+    fn on_event(&self, event: &Event<'_>, ctx: Context<'_, S>) {
+        self.inner.on_event(event, ctx.clone());
+        self.layer.on_event(event, ctx);
+    }
+
+    #[inline]
+    fn on_enter(&self, id: &span::Id, ctx: Context<'_, S>) {
+        self.inner.on_enter(id, ctx.clone());
+        self.layer.on_enter(id, ctx);
+    }
+
+    #[inline]
+    fn on_exit(&self, id: &span::Id, ctx: Context<'_, S>) {
+        self.inner.on_exit(id, ctx.clone());
+        self.layer.on_exit(id, ctx);
+    }
+
+    #[inline]
+    fn on_close(&self, id: span::Id, ctx: Context<'_, S>) {
+        self.inner.on_close(id.clone(), ctx.clone());
+        self.layer.on_close(id, ctx);
+    }
+
+    #[inline]
+    fn on_id_change(&self, old: &span::Id, new: &span::Id, ctx: Context<'_, S>) {
+        self.inner.on_id_change(old, new, ctx.clone());
+        self.layer.on_id_change(old, new, ctx);
+    }
+
+    #[doc(hidden)]
+    unsafe fn downcast_raw(&self, id: TypeId) -> Option<*const ()> {
+        match id {
+            // If downcasting to `Self`, return a pointer to `self`.
+            id if id == TypeId::of::<Self>() => Some(self as *const _ as *const ()),
+
+            // Oh, we're looking for per-layer filters!
+            //
+            // This should only happen if we are inside of another `Layered`,
+            // and it's trying to determine how it should combine `Interest`s
+            // and max level hints.
+            //
+            // In that case, we should be considered to be "per-layer filtered"
+            // if *both* the outer layer and the inner layer/subscriber have
+            // per-layer filters. Otherwise, we should be considered to *not* be
+            // per-layer filtered (even if one or the other has per layer
+            // filters). This is because *we* are expected to handle the
+            // `Interest`/level hint combining at this level, and will be
+            // propagating a non-PLF hint/interest.
+            //
+            // Yes, this rule *is* slightly counter-intuitive, but it's
+            // necessary due to a weird edge case that can occur when two
+            // `Layered`s where one side is per-layer filtered and the other
+            // isn't are `Layered` together to form a tree. If we didn't have
+            // this rule, we would actually end up *ignoring* `Interest`s from
+            // the non-per-layer-filtered layers, since both branches would
+            // claim to have PLF.
+            //
+            // If you don't understand this...that's fine, just don't mess with
+            // it. :)
```

Comment:
i love this passive-aggressive smiley that says "stay out of my territory"

Reply:
it wasn't intended to be passive aggressive, it was supposed to be reassuring! the intended message was "it's okay if this stuff is complicated, it won't be on the test"... ._.